### PR TITLE
overlord: support snapctl --halt|--poweroff in gadget install-device

### PR DESCRIPTION
**Pull request by pedronis**
This also needed work in daemon and state to introduce and support new restart types RestartSystem(Halt|Poweroff)Now.

TODO: a spread test would be good, probably a follow up

### CODE REVIEW

File: overlord/devicestate/devicestate_install_mode_test.go [hunk 1]
```diff
@@ -486,6 +493,53 @@ func (s *deviceMgrInstallModeSuite) TestInstallWithInstallDeviceHookExpTasks(c *
...
+	restore = devicestate.MockInstallRun(func(mod gadget.Model, gadgetRoot, kernelRoot, device string, options install.Options, _ gadget.ContentObserver) (*install.InstalledSystemSideData, error) {
+		return nil, nil
+	})
+	defer restore()
+
+	restore = hookstate.MockRunHook(func(ctx *hookstate.Context, tomb *tomb.Tomb) ([]byte, error) {
+		c.Assert(ctx.HookName(), Equals, "install-device")
+
+		// snapctl reboot --halt
+		_, _, err := ctlcmd.Run(ctx, []string{"reboot", arg}, 0)
```
Comment (anonymouse64) [inline overlord/devicestate/devicestate_install_mode_test.go:509]: to be clear, in this suite we have already mocked the reboot machinery such that this won't end up calling the real `reboot` command in response to this line of code calling `snapctl reboot <arg>` correct ?

**Review by anonymouse64 — APPROVED** [review]
lgtm, one question about the name of the task now, but it is probably fine as-is

File: overlord/devicestate/devicemgr.go [hunk 2]
```diff
@@ -949,20 +949,26 @@ func (m *DeviceManager) ensureInstalled() error {
...
 			// TODO: what's a reasonable timeout for the install-device hook?
 			Snap: model.Gadget(),
 			Hook: "install-device",
 		}
-		installDevice := hookstate.HookTask(m.state, summary, hooksup, nil)
+		installDevice = hookstate.HookTask(m.state, summary, hooksup, nil)
 		addTask(installDevice)
 	}
 
 	restartSystem := m.state.NewTask("restart-system-to-run-mode", i18n.G("Ensure next boot to run mode"))
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/devicestate/devicemgr.go:964]: I couldn't think of a reasonable name that wouldn't be too verbose to capture the new nuances but essentially restarting or setting up to restart into run mode is what this does, so the name seems okish as is

File: overlord/devicestate/devicestate_install_mode_test.go (hunk 1 above)
Comment (pedronis) [reply overlord/devicestate/devicestate_install_mode_test.go:509]: this simulates "snapctl reboot", the real reboot happens via daemon, which here is not wired up at all. What gets wired up is state.RequestRestart that,  you can see that in devicestate_test.go:

`s.o = overlord.MockWithStateAndRestartHandler(nil, func(req state.RestartType) ...`

and feeds into the suite restartRequests []state.RestartType  field

Comment (anonymouse64) [reply overlord/devicestate/devicestate_install_mode_test.go:509]: ah right I forgot that it doesn't go that far to daemon, thanks for explaining

File: overlord/devicestate/devicemgr.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (anonymouse64) [reply overlord/devicestate/devicemgr.go:964]: yeah agreed I think the name is okish given that I can't come up with a better one

**Review by mvo5 — APPROVED** [review]
This looks still good!

**Comment by mvo5** [comment]
The test failure:
```
2021-05-02T19:19:47.1767806Z ##[error]2021-05-02 19:19:47 Error executing google:ubuntu-core-16-64:tests/core/upgrade (may021906-022972) : 
2021-05-02T19:19:47.1800531Z -----
2021-05-02T19:19:47.1801145Z + TARGET_SNAP=core
2021-05-02T19:19:47.1802322Z + os.query is-core18
2021-05-02T19:19:47.1803307Z + '[' 0 '!=' 0 ']'
2021-05-02T19:19:47.1804043Z + awk '/^core / {print($3)}'
2021-05-02T19:19:47.1825839Z + snap list
2021-05-02T19:19:47.1826778Z + '[' -f curChg ']'
2021-05-02T19:19:47.1827282Z + case "$SPREAD_REBOOT" in
2021-05-02T19:19:47.1827790Z ++ cat prevBoot
2021-05-02T19:19:47.1829182Z + cmd='snap install --dangerous /var/lib/snapd/snaps/core_x1.snap'
2021-05-02T19:19:47.1830377Z ++ eval snap install --dangerous /var/lib/snapd/snaps/core_x1.snap --no-wait
2021-05-02T19:19:47.1831854Z +++ snap install --dangerous /var/lib/snapd/snaps/core_x1.snap --no-wait
2021-05-02T19:19:47.1832623Z + chg_id=9
2021-05-02T19:19:47.1841660Z + echo 9
2021-05-02T19:19:47.1842915Z + retry -n 50 --wait 1 sh -c 'journalctl -b -u snapd | MATCH "Waiting for system reboot"'
2021-05-02T19:19:47.1843745Z grep error: pattern not found, got:
...
```
looks real - this message is indeed no longer in the logs.

**Comment by pedronis** [comment]
mmh

> looks real - this message is indeed no longer in the logs.

@mvo5 at some point I was confused by unit tests into changing a string, but now I can change it back and they seem to pass. I think my last commit should fix this hopefully

**Pull request merged**